V0.23.0 notes (#2474)
PennyLaneAI/pennylane: doc/releases/changelog-0.23.0.md

:orphan:

# Release 0.23.0-dev (current release)

<h3>New features since last release</h3>

> <h4> Finite-shot circuit cutting ✂️</h4>

Author: `More practical circuit cutting`?

> <h4> Finite-shot circuit cutting ✂️</h4>

Reviewer: more powerful?

* The new `@qml.cut_circuit_mc` transform allows evaluating QNodes of
  `N`-wire circuits with `WireCut` operations and sample-based measurements

>   on devices with fewer than `N` wires.

Author: Rephrased.

  [(#2313)](https://github.com/PennyLaneAI/pennylane/pull/2313)
  [(#2321)](https://github.com/PennyLaneAI/pennylane/pull/2321)
  [(#2332)](https://github.com/PennyLaneAI/pennylane/pull/2332)
  [(#2358)](https://github.com/PennyLaneAI/pennylane/pull/2358)
  [(#2382)](https://github.com/PennyLaneAI/pennylane/pull/2382)
  [(#2399)](https://github.com/PennyLaneAI/pennylane/pull/2399)
  [(#2407)](https://github.com/PennyLaneAI/pennylane/pull/2407)
  [(#2444)](https://github.com/PennyLaneAI/pennylane/pull/2444)

  With these new additions, samples from the original circuit can be
  simulated using a Monte Carlo method, using fewer qubits at the expense of
  more device executions.  Additionally, this transform can take an optional
  classical processing function as an argument and return an expectation
  value.

  The following `3`-qubit circuit contains a `WireCut` operation and a `sample`
  measurement. When decorated with `@qml.cut_circuit_mc`, we can cut the circuit
  into two `2`-qubit fragments:

  ```python
  dev = qml.device("default.qubit", wires=2, shots=1000)

  @qml.cut_circuit_mc
  @qml.qnode(dev)

>   def circuit(x):

Author: Think so, though ran into an issue with drawing. Double-checking this with Tom and Ant.

>   def circuit(x):

Author: After more investigations, the bug here with drawing the circuit is, unfortunately, more intricate. We could target v0.24.0 for a fix or if more urgent, make a bug fix release.

      qml.RX(0.89, wires=0)
      qml.RY(0.5, wires=1)
      qml.RX(1.3, wires=2)

      qml.CNOT(wires=[0, 1])
      qml.WireCut(wires=1)
      qml.CNOT(wires=[1, 2])

      qml.RX(x, wires=0)
      qml.RY(0.7, wires=1)
      qml.RX(2.3, wires=2)
      return qml.sample(wires=[0, 2])
  ```

  we can then execute the circuit as usual by calling the QNode:

  ```pycon
  >>> x = 0.3
  >>> circuit(x)
  tensor([[1, 1],
          [0, 1],
          [0, 1],
          ...,
          [0, 1],
          [0, 1],
          [0, 1]], requires_grad=True)
  ```

  Furthermore, the number of shots can be temporarily altered when calling
  the QNode:

  ```pycon
  >>> results = circuit(x, shots=123)
  >>> results.shape
  (123, 2)
  ```

  Using the Monte Carlo approach of [Peng et. al](https://arxiv.org/abs/1904.00102), the
  `cut_circuit_mc` transform also supports returning sample-based expectation values of
  observables that are diagonal in the computational basis, as shown below for a `ZZ` measurement

>   on wires `0` and `2`:

Author: Moved the example to UsageDetails and refed the docs.

  ```python
  dev = qml.device("default.qubit", wires=2, shots=10000)

  def observable(bitstring):
      return (-1) ** np.sum(bitstring)

  @qml.cut_circuit_mc(classical_processing_fn=observable)
  @qml.qnode(dev)
  def circuit(x):
      qml.RX(0.89, wires=0)
      qml.RY(0.5, wires=1)
      qml.RX(1.3, wires=2)

      qml.CNOT(wires=[0, 1])
      qml.WireCut(wires=1)
      qml.CNOT(wires=[1, 2])

      qml.RX(x, wires=0)
      qml.RY(0.7, wires=1)
      qml.RX(2.3, wires=2)
      return qml.sample(wires=[0, 2])
  ```

  We can now approximate the expectation value of the observable using

  ```pycon
  >>> circuit(x)
  tensor(-0.776, requires_grad=True)
  ```

> * Added the automatic graph partitioning method `qcut.kahypar_cut()` for cutting

Author: Updated.

  [(#2173)](https://github.com/PennyLaneAI/pennylane/pull/2173)
  [(#2197)](https://github.com/PennyLaneAI/pennylane/pull/2197)
  [(#2362)](https://github.com/PennyLaneAI/pennylane/pull/2362)
  [(#2230)](https://github.com/PennyLaneAI/pennylane/pull/2230)
  [(#2199)](https://github.com/PennyLaneAI/pennylane/pull/2199)
  [(#2371)](https://github.com/PennyLaneAI/pennylane/pull/2371)
  [(#2272)](https://github.com/PennyLaneAI/pennylane/pull/2272)
  [(#2164)](https://github.com/PennyLaneAI/pennylane/pull/2164)
  [(#2316)](https://github.com/PennyLaneAI/pennylane/pull/2316)
  [(#2385)](https://github.com/PennyLaneAI/pennylane/pull/2385)
  [(#2372)](https://github.com/PennyLaneAI/pennylane/pull/2372)
  [(#2415)](https://github.com/PennyLaneAI/pennylane/pull/2415)
  [(#2426)](https://github.com/PennyLaneAI/pennylane/pull/2426)
  [(#2352)](https://github.com/PennyLaneAI/pennylane/pull/2352)
  [(#2420)](https://github.com/PennyLaneAI/pennylane/pull/2420)
  [(#2465)](https://github.com/PennyLaneAI/pennylane/pull/2465)
  [(#2454)](https://github.com/PennyLaneAI/pennylane/pull/2454)

  The :mod:`~.qchem` module provides access to a driver function :func:`~.molecular_hamiltonian`

>   to generate the electronic Hamiltonian in a single call. For example,

Reviewer: It might also be good to mentioned quite early on what users currently relying on the external pennylane-qchem package should do on upgrading.

  ```python
  symbols = ["H", "H"]
  geometry = np.array([[0., 0., -0.66140414], [0., 0., 0.66140414]])

>   hamiltonian, qubits = qml.qchem.molecular_hamiltonian(symbols, geometry)

Reviewer: @soranjh I can't quite tell what these two codeblocks are demonstrating. What is different here compared to `qml.qchem` usage in previous PL versions?

Reviewer: It might be good to explicitly highlight the differences (in code) here instead.

Reviewer: Same goes for the two code blocks below

  The following `3`-qubit circuit contains a `WireCut` operation and a `sample`
  measurement. When decorated with `@qml.cut_circuit_mc`, we can cut the circuit
  into two `2`-qubit fragments:

  ```python
  dev = qml.device("default.qubit", wires=2, shots=1000)

  @qml.cut_circuit_mc
  @qml.qnode(dev)
  def circuit(x):
      qml.RX(0.89, wires=0)
      qml.RY(0.5, wires=1)
      qml.RX(1.3, wires=2)

      qml.CNOT(wires=[0, 1])
      qml.WireCut(wires=1)
      qml.CNOT(wires=[1, 2])

      qml.RX(x, wires=0)
      qml.RY(0.7, wires=1)
      qml.RX(2.3, wires=2)
      return qml.sample(wires=[0, 2])
  ```

  we can then execute the circuit as usual by calling the QNode:

  ```pycon
  >>> x = 0.3
  >>> circuit(x)
  tensor([[1, 1],
          [0, 1],
          [0, 1],
          ...,
          [0, 1],
          [0, 1],
          [0, 1]], requires_grad=True)
  ```

  Furthermore, the number of shots can be temporarily altered when calling
  the QNode:

  ```pycon
  >>> results = circuit(x, shots=123)
  >>> results.shape
  (123, 2)
  ```

  The `cut_circuit_mc` transform also supports returning sample-based
  expectation values of observables using the `classical_processing_fn`

>   argument. Refer to the `UsageDetails` section of the [`cut_circuit_mc`
>   transform
>   documentation](https://pennylane.readthedocs.io/en/latest/code/api/pennylane.cut_circuit_mc.html)

Reviewer: ```suggestion
  argument. Refer to the Usage Details section of the [documentation](https://pennylane.readthedocs.io/en/latest/code/api/pennylane.cut_circuit_mc.html)
```

(since ReST doesn't support nested formatting :( )

![image](https://user-images.githubusercontent.com/2959003/165165977-ef6d61e6-5fe8-4f6c-a2fe-f58836590a63.png)

  for an example.

* The `cut_circuit` transform now supports automatic graph partitioning by
  specifying `auto_cutter=True` to cut arbitrary tape-converted graphs using
  the general purpose graph partitioning framework
  [KaHyPar](https://pypi.org/project/kahypar/).
  [(#2330)](https://github.com/PennyLaneAI/pennylane/pull/2330)
  [(#2428)](https://github.com/PennyLaneAI/pennylane/pull/2428)

  Note that `KaHyPar` needs to be installed separately with the `auto_cutter=True` option.

  For integration with the  existing low-level manual cut pipeline, refer to
  the [documentation of the `qcut.find_and_place_cuts`

>   function](https://pennylane.readthedocs.io/en/latest/code/api/pennylane.transforms.qcut.find_and_place_cuts.html)

Reviewer: ```suggestion
  the [documentation](https://pennylane.readthedocs.io/en/latest/code/api/pennylane.transforms.qcut.find_and_place_cuts.html).
```

>   `method='pyscf'` in :func:`~.molecular_hamiltonian`.

Reviewer: woah I had no idea we could use `:func:` inside the changelog (using markdown) and it would work!!!!

Reviewer: We should do this more to create links to docstrings, _but_ it does mean we have to strip this out before copying to the github release notes :(

  ```python
  symbols = ["H", "H"]
  geometry = np.array([[0.0, 0.0, 0.0], [0.0, 0.0, 2.0]])
  hamiltonian, qubits = qml.qchem.molecular_hamiltonian(symbols, geometry, method='pyscf')
  ```

>   Summary of QChem changes:

Reviewer: Some of these sounds like they should be proper, elaborated, changelog entries. In particular:

- New functions are added for computing multipole moment molecular integrals
- New functions are added for building a differentiable dipole moment observable
- New functions are added for building fermionic and qubit observables
- A new module is created for hosting openfermion to pennylane observable conversion functions
- The tapering functions are added to qchem
- The dependency on openbabel is removed
- The efficiency of computing molecular integrals and Hamiltonian is improved
- The 6-31G basis set is added to the qchem basis set repo

@soranjh I'm not sure the best place for them, but maybe either New Features or Improvements as you see appropriate?

They sound like _major_ changes that readers would want to know about!

I would remove the following, since these do not sound like important information for the release notes:

- Expressive names are used for the Hartree-Fock solver functions [This sounds like an implementation detail change?]
- Differentiable and non-differentiable backends can be selected for building a Hamiltonian [already mentioned]
- The quantum chemistry functionalities are unified [already mentioned]
- The qchem and new hf modules are merged [already mentioned]
- These new additions are added to a feature branch [not a change/feature 🙂]

  - New functions are added for computing multipole moment molecular integrals
  - New functions are added for building a differentiable dipole moment observable
  - External dependencies are replaced with local functions for spin and particle number observables
  - New functions are added for building fermionic and qubit observables
  - A new module is created for hosting openfermion to pennylane observable conversion functions
  - Expressive names are used for the Hartree-Fock solver functions
  - These new additions are added to a feature branch
  - The efficiency of computing molecular integrals and Hamiltonian is improved
  - The qchem and new hf modules are merged
  - The 6-31G basis set is added to the qchem basis set repo
  - The dependency on openbabel is removed
  - The tapering functions are added to qchem
  - Differentiable and non-differentiable backends can be selected for building a Hamiltonian
  - The quantum chemistry functionalities are unified

> <h4>Pattern matching optimization 🔎 💎 </h4>

Reviewer: @antalszava your call here, but you could merge the following subsections:

- New transforms for compilation and transpilation

  * Pattern matching
  * Swap-based transpiler

- A growing collection of templates

  * HilbertSchmidt and the LocalHilbertSchmidt 
  * qml.MERA

Author: That's a fair point, I was considering that. Then it came to mind that in this release we could afford to have sections be dedicated to a specific transform/template, so thought that the way they convey more information to the users. Having said that, happy to change as suggested, if we'd like that.

Reviewer: @antalszava completely up to you :) I don't mind either approach - feel free to leave it if you prefer

* Added an optimization transform that matches pieces of user-provided identity
  templates in a circuit and replaces them with an equivalent component.
  [(#2032)](https://github.com/PennyLaneAI/pennylane/pull/2032)

>   First let's consider the following circuit where we want to replace sequence of two ``pennylane.S`` gates with a

Reviewer: ```suggestion
  For example, consider the following circuit where we want to
  replace sequence of two `pennylane.S` gates with a
```

>   ``pennylane.PauliZ`` gate.

Reviewer: ```suggestion
  `pennylane.PauliZ` gate.
```

  ```python
  def circuit():
      qml.S(wires=0)
      qml.PauliZ(wires=0)
      qml.S(wires=1)
      qml.CZ(wires=[0, 1])
      qml.S(wires=1)
      qml.S(wires=2)
      qml.CZ(wires=[1, 2])
      qml.S(wires=2)
      return qml.expval(qml.PauliX(wires=0))
  ```

>   Therefore we use the following pattern that implements the identity:

Reviewer: ```suggestion
  We specify the following pattern that implements the identity:
```

  ```python
  with qml.tape.QuantumTape() as pattern:
      qml.S(wires=0)
      qml.S(wires=0)
      qml.PauliZ(wires=0)
  ```

>   For optimizing the circuit given the given following template of CNOTs we apply the `pattern_matching`
>   transform.

Reviewer: ```suggestion
  To optimize the circuit with this identity pattern, we apply the
  `qml.transforms.pattern_matching` transform.
```

  ```pycon
  >>> dev = qml.device('default.qubit', wires=5)
  >>> qnode = qml.QNode(circuit, dev)

>   >>> optimized_qfunc = qml.transforms.pattern_matching_optimization(pattern_tapes=[pattern])(circuit)

Reviewer: Is this correct? Should it be `pattern_matching` or `pattern_matching_optimization`?

Author: Yep, this one is `pattern_matching_optimization`.

  >>> optimized_qnode = qml.QNode(optimized_qfunc, dev)

> 
>   >>> print(qml.draw(qnode)())

Reviewer: ```suggestion
  >>> print(qml.draw(qnode)())
```

  0: ──S──Z─╭C──────────┤  <X>
  1: ──S────╰Z──S─╭C────┤

>   2: ──S──────────╰Z──S─┤
> 
>   >>> print(qml.draw(optimized_qnode)())

Reviewer: ```suggestion
  2: ──S──────────╰Z──S─┤
  >>> print(qml.draw(optimized_qnode)())
```

  0: ──S⁻¹─╭C────┤  <X>
  1: ──Z───╰Z─╭C─┤
  2: ──Z──────╰Z─┤
  ```

  For more details on using pattern matching optimization you can check the corresponding documentation and also the

>   following [paper](https://dl.acm.org/doi/full/10.1145/3498325).

Reviewer: Should link directly to the function docstring here!

<h4>Measure the distance between two unitaries📏</h4>

* Added the `HilbertSchmidt` and the
  `LocalHilbertSchmidt` templates to be used for computing distance measures
  between unitaries.
  [(#2364)](https://github.com/PennyLaneAI/pennylane/pull/2364)

  ```python
  with qml.tape.QuantumTape(do_queue=False) as u_tape:
      qml.Hadamard(wires=0)

  def v_function(params):

>       qml.RZ(params[0], wires=1)

Reviewer: I think there should be some text in-between these function/tape definitions, explaining what they are/represent

Reviewer: Actually, I'm curious how come one is represented as a tape, and the other as a function? Why not define both using quantum functions?

```python
  def u_function():
      qml.Hadamard(wires=0)

  def v_function(params):
      qml.RZ(params[0], wires=1)
```

Author: Added small comments here. Otherwise as discussed, this would be a matter of more thinking.

  @qml.qnode(dev)
  def hilbert_test(v_params, v_function, v_wires, u_tape):
      qml.HilbertSchmidt(v_params, v_function=v_function, v_wires=v_wires, u_tape=u_tape)
      return qml.probs(u_tape.wires + v_wires)

  def cost_hst(parameters, v_function, v_wires, u_tape):
      return (1 - hilbert_test(v_params=parameters, v_function=v_function, v_wires=v_wires, u_tape=u_tape)[0])
  ```
  ```pycon
  >>> cost_hst(parameters=[0.1], v_function=v_function, v_wires=[1], u_tape=u_tape)
  tensor(0.999, requires_grad=True)

>   ```

Reviewer: Could be good to briefly mention what this quantity is, and to provide a link to the docstrings!

  `LocalHilbertSchmidt` templates to be used for computing distance measures
  between unitaries.
  [(#2364)](https://github.com/PennyLaneAI/pennylane/pull/2364)

  Given a unitary `U`, `qml.HilberSchmidt` can be used to measure the distance
  between unitaries and to define a cost function (`cost_hst`) used for
  learning a unitary `V` that is equivalent to `U` up to a global phase:
  ```python
  # Represents unitary U
  with qml.tape.QuantumTape(do_queue=False) as u_tape:
      qml.Hadamard(wires=0)

  # Represents unitary V
  def v_function(params):
      qml.RZ(params[0], wires=1)

  @qml.qnode(dev)
  def hilbert_test(v_params, v_function, v_wires, u_tape):
      qml.HilbertSchmidt(v_params, v_function=v_function, v_wires=v_wires, u_tape=u_tape)
      return qml.probs(u_tape.wires + v_wires)

  def cost_hst(parameters, v_function, v_wires, u_tape):
      return (1 - hilbert_test(v_params=parameters, v_function=v_function, v_wires=v_wires, u_tape=u_tape)[0])
  ```
  ```pycon
  >>> cost_hst(parameters=[0.1], v_function=v_function, v_wires=[1], u_tape=u_tape)
  tensor(0.999, requires_grad=True)
  ```
  For more information refer to the [documentation of

>   qml.HilbertSchmidt](https://pennylane.readthedocs.io/en/latest/code/api/pennylane.HilbertSchmidt.html?highlight=hilbertschmidt#pennylane.HilbertSchmidt).

Reviewer: ```suggestion
  qml.HilbertSchmidt](https://pennylane.readthedocs.io/en/latest/code/api/pennylane.HilbertSchmidt.html).
```

  ```
  It may be necessary to reorder the wires to see the MERA architecture clearly:
  ```pycon
  >>> print(qml.draw(circuit,expansion_strategy='device',wire_order=[2,0,1,3])(template_weights))
  2: ───────────────╭C──RY(0.10)──╭X──RY(-0.30)───────────────┤
  0: ─╭X──RY(-0.30)─│─────────────╰C──RY(0.10)──╭C──RY(0.10)──┤
  1: ─╰C──RY(0.10)──│─────────────╭X──RY(-0.30)─╰X──RY(-0.30)─┤  <Z>
  3: ───────────────╰X──RY(-0.30)─╰C──RY(0.10)────────────────┤
  ```

<h4>New transform for transpilation ⚙️ </h4>

* Added a swap based transpiler transform.
  [(#2118)](https://github.com/PennyLaneAI/pennylane/pull/2118)

  The transpile function takes a quantum function and a coupling map as inputs
  and compiles the circuit to ensure that it can be executed on corresponding
  hardware. The transform can be used as a decorator in the following way:

  ```python
  dev = qml.device('default.qubit', wires=4)

  @qml.qnode(dev)
  @qml.transforms.transpile(coupling_map=[(0, 1), (1, 2), (2, 3)])
  def circuit(param):
      qml.CNOT(wires=[0, 1])
      qml.CNOT(wires=[0, 2])
      qml.CNOT(wires=[0, 3])
      qml.PhaseShift(param, wires=0)
      return qml.probs(wires=[0, 1, 2, 3])

>   ```

Reviewer: can we include output here? For example, might be good to include a circuit drawing showing the compiled circuit

<h3>Improvements</h3>

> * `QuantumTape` objects are now iterable and accessing the
>   operations and measurements of the underlying quantum circuit is more

Reviewer: ```suggestion
* `QuantumTape` objects are now iterable, allowing iteration over the
  contained operations and measurements.
```

   expval(PauliZ(wires=[0]))]
  ```

> * Added the `QuantumTape.shape` method and `QuantumTape.numeric_type`
>   attribute to allow extracting information about the shape and numeric type of
>   quantum tapes.
>   [(#2044)](https://github.com/PennyLaneAI/pennylane/pull/2044)

Reviewer: _may_ be worth adding a _very_ short (*very* short) example  here!
```suggestion
* Added the `QuantumTape.shape` method and `QuantumTape.numeric_type`
  attribute to allow extracting information about the shape and numeric type of
  the output returned by a quantum tape after execution.
  [(#2044)](https://github.com/PennyLaneAI/pennylane/pull/2044)
```

* Defined a `MeasurementProcess.shape` method and a
  `MeasurementProcess.numeric_type` attribute.

>   [(#2044)](https://github.com/PennyLaneAI/pennylane/pull/2044)

Reviewer: This should specify _what_ these attributes allow you to extract! Like in the above docstring.

* The parameter-shift Hessian can now be computed for arbitrary
  operations that support the general parameter-shift rule for
  gradients, using `qml.gradients.param_shift_hessian`
  [(#2319)](https://github.com/XanaduAI/pennylane/pull/2319)

  Multiple ways to obtain the
  gradient recipe are supported, in the following order of preference:

  - A custom `grad_recipe`. It is iterated to obtain the shift rule for
    the second-order derivatives in the diagonal entries of the Hessian.

  - Custom `parameter_frequencies`. The second-order shift rule can
    directly be computed using them.

  - An operation's `generator`. Its eigenvalues will be used to obtain
    `parameter_frequencies`, if they are not given explicitly for an operation.

* The `qml.specs` transform now accepts an `expansion_strategy` keyword argument.

>   [(#2395)](https://github.com/PennyLaneAI/pennylane/pull/2395)

Reviewer: This should also describe what the change actually means :)

> * `default.qubit` and `default.mixed` now skip over identity operators instead of performing matrix multiplication

Reviewer: ```suggestion
* The `default.qubit` and `default.mixed` devices now skip over identity operators instead of performing matrix multiplication
```

>   with the identity.

Reviewer: 😆

  accessed via the top-level `qml` namespace.
  [(#2396)](https://github.com/PennyLaneAI/pennylane/pull/2396)

<h3>Breaking changes</h3>

> * The caching ability of `QubitDevice` has been removed, using the caching on

Reviewer: ```suggestion
* The caching ability of devices has been removed. Using the caching on
```

  the QNode level is the recommended alternative going forward.

>   [(#2443)](https://github.com/PennyLaneAI/pennylane/pull/2443)

Reviewer: Very nice how you have an example showing what a user should do instead!

  1
  ```

> * The `qml.finite_diff` function has been deleted. Please use `qml.gradients.finite_diff` to compute

Reviewer: ```suggestion
* The `qml.finite_diff` function has been removed. Please use `qml.gradients.finite_diff` to compute
```

  the gradient of tapes of QNodes. Otherwise, manual implementation is required.
  [(#2464)](https://github.com/PennyLaneAI/pennylane/pull/2464)

> * The `get_unitary_matrix` transform has been removed, users should use
>   `qml.matrix` instead.
>   [(#2457)](https://github.com/PennyLaneAI/pennylane/pull/2457)

Reviewer: ```suggestion
* The `get_unitary_matrix` transform has been removed, please use
  `qml.matrix` instead.
  [(#2457)](https://github.com/PennyLaneAI/pennylane/pull/2457)
```

I don't like having the term 'user' appear in the changelog 😆

Author: Makes sense, I guess we are talking to users through the release notes, it feels a bit weird to mention them in third person.

Author: Haven't found other instances.

Reviewer: > I guess we are talking to users through the release notes, it feels a bit weird to mention them in third person.

💯

> * The `update_stepsize` method is being deleted from `GradientDescentOptimizer` and its child

Reviewer: ```suggestion
* The `update_stepsize` method has been removed from `GradientDescentOptimizer` and its child
```

In case it happens elsewhere and I didn't spot it, lets make sure that:

- We are using the past tense, not present tense
- We use 'removed' instead of 'deleted'.

Author: :+1: For my understanding: what would be the difference between removed and deleted?

Reviewer: Oh it just sounds more formal in my head!

>   - `qml.drawer.CircuitDrawer` is replaced by `qml.drawer.tape_text`.

Reviewer: On line 516, 'is being deleted.' -> 'has been removed'

  - `qml.drawer.CHARSETS` is deleted because we now assume everyone has access to unicode.
  - `Grid` and `qml.drawer.drawable_grid` are removed because the custom data class is replaced
    by list of sets of operators or measurements.

>   - `RepresentationResolver` is replaced by the `Operator.label` method.

Reviewer: Of all these changes, I think `Operator.label` is the one that should be elaborated on and is most important to developers.

Author: Moved it to the first spot.

<h3>Bug fixes</h3>

* Fixed a bug arising when multiplying `Hamiltonian` objects.

>   [(#2466)](https://github.com/PennyLaneAI/pennylane/pull/2466)

Reviewer: Not essential, but might be nice to briefly mention what the bug was.

Reviewer: It can still be vague, e.g., like the one below:

> Fixed a bug in the qml.PauliRot operation, where computing the generator was not taking into account the operation wires. 

It just allows readers to work out if they might have been affected by this bug, and whether they should upgrade to fix it.

* Fixed a bug where non-trainable arguments were shifted in the `NesterovMomentumOptimizer`
  if a trainable argument was after it in the argument list.
  [(#2466)](https://github.com/PennyLaneAI/pennylane/pull/2466)

* Fixed a bug with `@jax.jit` for grad when `diff_method="adjoint"` and `mode="backward"`.
  [(#2460)](https://github.com/PennyLaneAI/pennylane/pull/2460)

* Fixed a bug where `qml.DiagonalQubitUnitary` did not support `@jax.jit`
  and `@tf.function`.
  [(#2445)](https://github.com/PennyLaneAI/pennylane/pull/2445)

* Fixed a bug in the `qml.PauliRot` operation, where computing the generator was not taking into
  account the operation wires.
  [(#2442)](https://github.com/PennyLaneAI/pennylane/pull/2442)

* Fixed a bug with the padding capability of `AmplitudeEmbedding` where the
  inputs are on the GPU.
  [(#2431)](https://github.com/PennyLaneAI/pennylane/pull/2431)

* Fixed a bug by adding a comprehensible error message for calling `qml.probs`
  without passing wires or an observable.
  [(#2438)](https://github.com/PennyLaneAI/pennylane/pull/2438)

> * Call `pip show` with the subprocess library to avoid outputting a common warning.
>   [(#2422)](https://github.com/PennyLaneAI/pennylane/pull/2422)

Reviewer: The wording on this one doesn't make sense?
```suggestion
* The behaviour of `qml.about()` was modified to avoid warnings being emitted
  due to legacy behaviour of `pip`.
  [(#2422)](https://github.com/PennyLaneAI/pennylane/pull/2422)
```

* Optimizers only consider a variable trainable if they have `requires_grad = True`.
  [(#2381)](https://github.com/PennyLaneAI/pennylane/pull/2381)

* Fixed a bug with `qml.expval`, `qml.var`, `qml.state` and
  `qml.probs` (when `qml.probs` is the only measurement) where the `dtype`
  specified on the device did not match the `dtype` of the QNode output.
  [(#2367)](https://github.com/PennyLaneAI/pennylane/pull/2367)

* Fixed a bug where the output shapes from batch transforms are inconsistent
  with the QNode output shape.
  [(#2215)](https://github.com/PennyLaneAI/pennylane/pull/2215)

* Fixed a bug caused by the squeezing in `qml.gradients.param_shift_hessian`.
  [(#2215)](https://github.com/PennyLaneAI/pennylane/pull/2215)

* Fixed a bug in which the `expval`/`var` of a `Tensor(Observable)` would depend on the order
  in which the observable is defined:
  [(#2276)](https://github.com/PennyLaneAI/pennylane/pull/2276)
  ```pycon

>   @qml.qnode(dev)

Reviewer: @antalszava you'll need to write the entire block in pycon syntax:

````markdown
```pycon
>>> @qml.qnode(dev)
... def circ(op):
...   qml.RX(0.12, wires=0)
...   qml.RX(1.34, wires=1)
...   qml.RX(3.67, wires=2)
...   return qml.expval(op)
>>> op1 = qml.Identity(wires=0) @ qml.Identity(wires=1) @ qml.PauliZ(wires=2)
>>> op2 = qml.PauliZ(wires=2) @ qml.Identity(wires=0) @ qml.Identity(wires=1)
>>> print(circ(op1), circ(op2))
-0.8636111153905662 -0.8636111153905662
```
````

  ```

>   [(#2276)](https://github.com/PennyLaneAI/pennylane/pull/2276)

Reviewer: this PR link should be moved to the end of the line with the bullet point!

Reviewer: Also I think we can convert the two codeblocks above into a single `pycon` code block

* Fixed a bug where `qml.hf.transform_hf()` would fail due to missing wires in
  the qubit operator that is prepared for tapering the HF state.
  [(#2441)](https://github.com/PennyLaneAI/pennylane/pull/2441)

<h3>Documentation</h3>

* The sections on adding operator and observable support in the "How to add a
  plugin" section of the plugins page have been updated.
  [(#2389)](https://github.com/PennyLaneAI/pennylane/pull/2389)

> * Added the missing arxiv reference in `LieAlgebra` optimizer.
>   [(#2325)](https://github.com/PennyLaneAI/pennylane/pull/2325)

Reviewer: ```suggestion
* The missing arXiv reference in the `LieAlgebra` optimizer has been fixed.
  [(#2325)](https://github.com/PennyLaneAI/pennylane/pull/2325)
```
